### PR TITLE
Refactor loop metadata handling

### PR DESCRIPTION
**Pull request by tcorringham**
Moved the loop metadata handling from llpc/lower to lgc/patch as a first step in extending the functionality to allow more flexible handling of loop metadata, e.g. DontUnroll.
This change preserves the previous behavior.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit c5863ec78f501a6d02da04d8c27974c72eb33e1c
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev a3a42838c576cc219e61500bc469a4a05ce0db68</li>
<li>SPVGEN: dev a25eab81bf58bc6a95e178d44db16c48e8bf7da1</li>
<li>XGL: dev c7b65e3a8497318edd4ff1c290fbfa47398090d7</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7ce7cdb9b22229245c7c0e836584561baeb8f776</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 979/979 (100.0%)</li>
<li>Failed: 0/979 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142137/142137 (100.0%)</li>
<li>Failed: 0/142137 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142042/142043 (100.0%)</li>
<li><details><summary>Failed: 1/142043 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.spirv_assembly.type.vec3.u64.shift_right_arithmetic_shift8_vert
Stack trace: Flake
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 8.2, Ellesmere</summary><ul>

<li>Passed: 142043/142043 (100.0%)</li>
<li>Failed: 0/142043 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 93bc2e6f947e6fc52fae96af17ac78f4b0297efd
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev a3a42838c576cc219e61500bc469a4a05ce0db68</li>
<li>SPVGEN: dev a25eab81bf58bc6a95e178d44db16c48e8bf7da1</li>
<li>XGL: dev c7b65e3a8497318edd4ff1c290fbfa47398090d7</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7ce7cdb9b22229245c7c0e836584561baeb8f776</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 979/979 (100.0%)</li>
<li>Failed: 0/979 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142137/142137 (100.0%)</li>
<li>Failed: 0/142137 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142043/142043 (100.0%)</li>
<li>Failed: 0/142043 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Ellesmere</summary><ul>

<li>Passed: 142043/142043 (100.0%)</li>
<li>Failed: 0/142043 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
Hi @tcorringham,
What's the bigger picture? In the long term, is this to allow smarter handling of loop annotations, e.g., treating annotations like DontUnroll as hints to the optimizer?

**Comment by tcorringham** [comment]
Yes, this is the first step in being able to treat hints such as DontUnroll in more flexible ways. Currently the handling of hints is rather rigid, which is not appropriate in all situations.

**Comment by kuhar** [comment]
SGTM. Could you update the PR description and the commit message to mention that?

**Comment by amdvlk-admin** [comment]
### Test summary for commit dfa69969efa268993301c028ec20d91499f5b3c3
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev a3a42838c576cc219e61500bc469a4a05ce0db68</li>
<li>SPVGEN: dev a25eab81bf58bc6a95e178d44db16c48e8bf7da1</li>
<li>XGL: dev c7b65e3a8497318edd4ff1c290fbfa47398090d7</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7ce7cdb9b22229245c7c0e836584561baeb8f776</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 979/979 (100.0%)</li>
<li>Failed: 0/979 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142137/142137 (100.0%)</li>
<li>Failed: 0/142137 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142043/142043 (100.0%)</li>
<li>Failed: 0/142043 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Ellesmere</summary><ul>

<li>Passed: 142043/142043 (100.0%)</li>
<li>Failed: 0/142043 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
can one of the code reviewers verify this patch?

**Comment by JaxLinAMD** [comment]
is there any other code reviewer who want to take a look? @GPUOpen-Drivers/llpcreviewers

File: lgc/patch/PatchLoopMetadata.cpp [hunk 1]
```diff
@@ -24,91 +24,86 @@
...
-  if (m_context->getPipelineContext()) {
-    auto shaderOptions = &(m_context->getPipelineShaderInfo(m_shaderStage)->options);
-    if (shaderOptions->forceLoopUnrollCount > 0)
+bool PatchLoopMetadata::runOnModule(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass patch-loop-metadata\n");
+
+  Patch::init(&module);
+  PipelineState *m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(&module);
+  m_gfxIp = m_pipelineState->getTargetInfo().getGfxIpVersion();
+  bool Changed = false;
```
Comment (kuhar) [inline lgc/patch/PatchLoopMetadata.cpp:94]: nit: changed

File: lgc/patch/PatchLoopMetadata.cpp [hunk 2]
```diff
@@ -24,91 +24,86 @@
...
+
+  for (Function &func : module) {
+    ShaderStage stage = getShaderStage(&func);
+    if (stage == -1)
+      continue;
+    auto shaderOptions = &m_pipelineState->getShaderOptions(stage);
+    if (shaderOptions) {
```
Comment (kuhar) [inline lgc/patch/PatchLoopMetadata.cpp:101]: nit: you can fold these two and put the definition in the if condition if it's not used elsewhere, i.e.:
` if (auto shaderOptions = &...)`

File: lgc/interface/lgc/Pipeline.h [hunk 3]
```diff
@@ -164,6 +164,15 @@ struct ShaderOptions {
 
   /// Default unroll threshold for LLVM.
   unsigned unrollThreshold;
+
+  /// Unroll loops by specified amount. 0 is default, 1 is no unroll.
```
Comment (amdrexu) [inline lgc/interface/lgc/Pipeline.h:168]: Please change /// to //. I decide to clean up all unnecessary style of "///". Some are introduced by LGC separation which extract certain interface definitions from llpc.h.

File: lgc/patch/PatchLoopMetadata.cpp (hunk 1 above)
Comment (amdrexu) [reply lgc/patch/PatchLoopMetadata.cpp:94]: The same as Jakub. Variable is lowercase-initial, different from LLVM style.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 57f5df62ac79b479da8fbe22e297c203b616082b
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev a3a42838c576cc219e61500bc469a4a05ce0db68</li>
<li>SPVGEN: dev a25eab81bf58bc6a95e178d44db16c48e8bf7da1</li>
<li>XGL: dev c7b65e3a8497318edd4ff1c290fbfa47398090d7</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142137/142137 (100.0%)</li>
<li>Failed: 0/142137 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142043/142043 (100.0%)</li>
<li>Failed: 0/142043 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142087/142087 (100.0%)</li>
<li>Failed: 0/142087 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/interface/lgc/Pipeline.h (hunk 3 above)
Comment (tcorringham) [reply lgc/interface/lgc/Pipeline.h]: Done

File: lgc/patch/PatchLoopMetadata.cpp (hunk 1 above)
Comment (tcorringham) [reply lgc/patch/PatchLoopMetadata.cpp]: Done

File: lgc/patch/PatchLoopMetadata.cpp (hunk 2 above)
Comment (tcorringham) [reply lgc/patch/PatchLoopMetadata.cpp]: Done

**Comment by JaxLinAMD** [comment]
@tcorringham please resolve conflicts

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7b41acb0c93bd1cdf6103b693205694754a87c50
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev d128e79bf98557761304e6aa46ba143fffb551d8</li>
<li>SPVGEN: dev 6205d0b42ebec5c8c01ff41e97fe51b56aa67343</li>
<li>XGL: dev d34fcbf7a0a5755cbac7b27c5784b7ed6782e4c2</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 142137/142137 (100.0%)</li>
<li>Failed: 0/142137 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 142043/142043 (100.0%)</li>
<li>Failed: 0/142043 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 142087/142087 (100.0%)</li>
<li>Failed: 0/142087 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@tcorringham uh, conficts again

**Comment by JaxLinAMD** [comment]
@tcorringham, please resolve conflicts.

**Comment by JaxLinAMD** [comment]
@tcorringham ping.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f41dc10daf39c4738589ce8b60ba92379a02a517
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev b30561984a7b697ed7c709be76ff2e4d30532728</li>
<li>SPVGEN: dev b28e58f562591f335b9a3b5ae7e578a426df210b</li>
<li>XGL: dev 9bcc23fb29dab0b7ed1ebff17f3e99ae4da0bdf9</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/441732)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 147198/147198 (100.0%)</li>
<li>Failed: 0/147198 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 147242/147242 (100.0%)</li>
<li>Failed: 0/147242 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 147292/147292 (100.0%)</li>
<li>Failed: 0/147292 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@tcorringham  could you resolve conficts? you need to push the commit

**Comment by tcorringham** [comment]
I don't see any conflicts with the base branch. It needs somebody with write access to merge the pull request - I don't have write access.

**Comment by amdrexu** [comment]
> I don't see any conflicts with the base branch. It needs somebody with write access to merge the pull request

Maybe it doesn't show in your view. You can rebase your change by fetching latest codes from dev branch with git pull --rebase. Then, git will give conflict message and requests you to resolve the conflicts.

**Comment by perlfu** [comment]
> > I don't see any conflicts with the base branch. It needs somebody with write access to merge the pull request
> 
> Maybe it doesn't show in your view. You can rebase your change by fetching latest codes from dev branch with git pull --rebase. Then, git will give conflict message and requests you to resolve the conflicts.

GitHub does seem to show conflicts that don't exist.  You have to do a rebase locally and force push that back to the pull request branch to clear them.

**Comment by dstutt** [comment]
I don't see conflicts at this point. I've got the "Merge pull request" button and an indication that it should be possible to merge automatically.

**Comment by dstutt** [comment]
Jax - do you want me to merge the change?

**Comment by JaxLinAMD** [comment]
@dstutt the button you are seeing is "Create a merge commit", I am trying to use "Rebase and merge", if you select "Rebase and merge" in the drop down list, you can also see the conflicts

**Comment by JaxLinAMD** [comment]
in fact, merge is also Ok if rebase is too complex.

**Comment by dstutt** [comment]
The button I see is "Merge pull request"

**Comment by JaxLinAMD** [comment]
I will merge it with "Create a merge commit" instead of "Rebase and merge"

**Comment by dstutt** [comment]
Oh - hang on, I see what you mean now. Yes I get conflicts with the rebase and merge option

**Comment by tcorringham** [comment]
Is this perhaps caused by me resolving the earlier trivial conflicts using the on-line resolve conflicts feature rather than pushing an update? If so I won't use that again!

**Comment by JaxLinAMD** [comment]
yes.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f4e0478d1a7c9642e90679e1498752b323c8c071
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev b30561984a7b697ed7c709be76ff2e4d30532728</li>
<li>SPVGEN: dev b28e58f562591f335b9a3b5ae7e578a426df210b</li>
<li>XGL: dev 9bcc23fb29dab0b7ed1ebff17f3e99ae4da0bdf9</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/430320)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 143394/143394 (100.0%)</li>
<li>Failed: 0/143394 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 143438/143438 (100.0%)</li>
<li>Failed: 0/143438 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 143488/143488 (100.0%)</li>
<li>Failed: 0/143488 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@tcorringham there are new conflicts, please resolve it.

**Comment by amdvlk-admin** [comment]
### Test summary for commit f7a1c21fcadbbaa34d58822a2713b5451cb297b0
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev b30561984a7b697ed7c709be76ff2e4d30532728</li>
<li>SPVGEN: dev b28e58f562591f335b9a3b5ae7e578a426df210b</li>
<li>XGL: dev 9bcc23fb29dab0b7ed1ebff17f3e99ae4da0bdf9</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev b3a0782762d195b57a2bebaabca24be90a35db2a</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/430320)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 143394/143394 (100.0%)</li>
<li>Failed: 0/143394 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 143438/143438 (100.0%)</li>
<li>Failed: 0/143438 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 143487/143488 (100.0%)</li>
<li><details><summary>Failed: 1/143488 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.info.vulkan1p2.feature_bits_influence
Stack trace: Flake
```
</p></details></li>


</ul></details>


</ul></details>

**Comment by jayfoad** [comment]
This seems to be causing llpc lit test failures for me:
```
Failed Tests (3):
  LLPC_SHADERTEST :: shaderdb/OpEmitStreamVertex_TestGeneral_lit.geom
  LLPC_SHADERTEST :: shaderdb/OpEmitVertex_TestGeneral_lit.geom
  LLPC_SHADERTEST :: shaderdb/OpImageFetch_TestDynamicOffset.spvasm


Testing Time: 25.59s
  Unsupported      :   1
  Passed           : 976
  Expectedly Failed:   3
  Failed           :   3
FAILED: compiler/llpc/test/CMakeFiles/check-amdllpc 
```

**Comment by kuhar** [comment]
@jayfoad what is your build configuration? The CI doesn't complain about this

**Comment by jayfoad** [comment]
Yeah, I noticed that and it's confusing. I'm using a cmake DEBUG build.

**Comment by dstutt** [comment]
It reproduces for me locally. I'm trying to work out why the CI isn't failing (internal CI is also passing).

**Comment by dstutt** [comment]
[AMD Official Use Only - Internal Distribution Only]

I think we've narrowed this down to an issue when the patch is merged back into our internal repository.

David
________________________________

**Pull request merged**